homeassistant/components/packages.py

```diff
@@ -0,0 +1,76 @@
+"""
+Support for Packages.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/packages/
+"""
+import logging
+
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'packages'
+CONF_PACKAGES = DOMAIN
+
+_COMPONENT_SCHEMA = vol.Schema({cv.slug: vol.Any(dict, list)})
+
+_CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({cv.slug: _COMPONENT_SCHEMA})
+}, extra=vol.ALLOW_EXTRA)
+
+
+def _log_error(package, component, message, *message_args):
+    """Log an error while merging."""
+    _LOGGER.error("Package %s setup failed. Component %s %s",
+                  package, component, message.format(*message_args))
+
+
+def merge_packages_config(config):
```

Comment:
Should merging take into consideration if a component is platform based? (can test by checking if component has `PLATFORM_CONFIG` ).  When platform based, dicts should not be merged but instead converted to a list that contains both dicts.

As you can see, we're getting into a super confusing area now in which the user is not able to distinguish where a piece of info came from.

Because what happens if you have in main config 

```yaml
switch:
  platform: bla

switch other:
  platform: other
```

and merge it with:

```yaml
switch:
  platform: merged
```

It will then become:

```yaml
switch:
 - platform: bla
 - platform: merged

switch other:
 - platform: other
```

which could influence the order entity ids are being generated.